0001-job-worker.md

```diff
@@ -0,0 +1,257 @@
+>   Job Status
+>	ID          667752ba-4cbb-44eb-adad-5b324c8204bc
+>	Owner       client1
+>	PID         387
+>	Running     true
+>	ExitCode    
+
+worker stream 667752ba-4cbb-44eb-adad-5b324c8204bc
+> hello
+> hello
+Ctrl+C
+
+worker stop 667752ba-4cbb-44eb-adad-5b324c8204bc
+
+worker status 667752ba-4cbb-44eb-adad-5b324c8204bc
+>   Job Status
+>	ID          667752ba-4cbb-44eb-adad-5b324c8204bc
+>	Owner       client1
+>	PID         387
+>	Running     false
+>	ExitCode    0
+```
+
+### Golang Usage
+
+Please see `doc.go` for an annotated list of types and interfaces the go library will provide. Below is a simple example of how the library could be used to start a job and stream the output
+
+```golang
+package main
+
+import (
+    "log"
+    "bufio"
+    "os/user"
+    "pkg/jobworker"
+)
+
+func main() {
+    worker := jobworker.New()
+
+    cmd := "while true; do echo hello; sleep 2; done"
+
+    // In the gRPC server the user ID would be extracted from the common name of the client tls cert
+    // For this example we simply use the current linux user
+    user, err := user.Current()
+    if err != nil {
+        log.Error(err)
+        return
+    }
+
+    id, err := worker.Start(user.UserName, JobOpts{100, "100M", "50ms"}, cmd)
+    if err != nil {
+        log.Error(err)
+        return
+    }
+
+    status, err := worker.Status(id)
+    if err != nil {
+        log.Error(err)
+        return
+    }
+
+    if !status.Running {
+        log.Error("job not running")
+        return
+    }
+
+    reader, err := worker.Output(id)
+    if err != nil {
+        log.Error("could not get reader for job's output")
+        return
+    }
+
+    scanner := bufio.NewScanner(reader)
+    for scanner.Scan() {
+        line := scanner.Text()
+        log.Printf("%s\n", line)
+    }
+}
+```
+
+### Audit Events
+
+Since the jobworker could be executing linux commands on behalf of a client some events would be useful for auditing purposes, similar to what you would expect with a production SSH server. Although not implemented in this example for brevity, the type of events emitted in a production system could be:
+
+ - start
+ - stop
+
+along with the "owner" and "command" that they executed. This can be useful not only to know what clients are executing but also counting how many commands a client is executing or determine the most common commands. These type of events could also help detect DDOS attacks or identify unknown / blacklisted clients.
+
+### Observability
+
+Since the jobworker would expose a gRPC server, in production this server should expose a health endpoint for monitoring purposes. In addition to this a metric such as "running jobs" would be useful to know how many jobs the jobworker is currently running. This could be used to trigger warnings if too many commands are seen (identifying a DDOS attack), too many clients, or new clients, are executing commands over a window of time etc.
+
+### Product Usage
```

Comment:
Have you thought about how you might test that the cgroup limits are correctly enforced?

Reply:
Each cgroup controller has an interface file for providing the resource usage of the controller. For CPU this is `cpu.stat`, memory is `mem.current` and IO is `io.stat`. Testing can be done by creating a cgroup and using a command that would exceed the cgroup's resource control and ensure only the limits are being used.

For example testing memory could be achieved by starting a job that uses a lot of memory, such as
`dd if=/dev/urandom of=/dev/shm/sample.txt bs=1G count=4 iflag=fullblock`
with job options limiting the memory to 100mb. While the job is being executed the `mem.current` should not exceed 100.

Another example for CPU could be a job with command `dd if=/dev/zero of=/dev/null`, again we specify the job options to limit CPU to say 100. CPU is a little more tricky since the quota is not a max, but we could compare `nr_periods` with `nr_throttled` and ensure the job is being throttled and at a ratio we would expect.

IO similar to CPU is a bit harder to test, since there is no hard limit and throttling is based on other processes IO. Here we could create two jobs with the same command as above. The first job will have an `io.latency` half that of the second. When comparing `io.stat` the first job should have more bytes written as the cgroup controller will prioritise the lowest latency.